libsolidity/codegen/ArrayUtils.cpp

```diff
@@ -823,6 +823,86 @@ void ArrayUtils::incrementDynamicArraySize(ArrayType const& _type) const
 		})", {"ref"});
 }
 
+void ArrayUtils::popStorageArrayElement(ArrayType const& _type) const
+{
+	solAssert(_type.location() == DataLocation::Storage, "");
+	solAssert(_type.isDynamicallySized(), "");
+	if (!_type.isByteArray() && _type.baseType()->storageBytes() < 32)
+		solAssert(_type.baseType()->isValueType(), "Invalid storage size for non-value type.");
+
+	if (_type.isByteArray())
+	{
+		m_context.appendInlineAssembly(R"({
+			let slot_value := sload(ref)
+			switch and(slot_value, 1)
+			case 0 {
+				// short byte array
+				let length := and(div(slot_value, 2), 0x1f)
+				if iszero(length) { invalid() }
+
+				// Zero-out the suffix including the least significant byte.
+				let mask := sub(exp(0x100, sub(33, length)), 1)
+				length := sub(length, 1)
+				slot_value := or(and(not(mask), slot_value), mul(length, 2))
+				sstore(ref, slot_value)
+			}
+			case 1 {
+				// long byte array
+				let length := div(slot_value, 2)
+				let slot := keccak256(0, 0x20)
```

Comment:
This and the next line have to be swapped. Why don't we have a test failure for this?

Can you please modify the tests such that the arrays are not the first element in storage, i.e. add some unused storage variables at the beginning of the contract.

Reply:
@chriseth I moved ``mstore(0, ref)`` in front of the ``keccak256`` call. But also without this change I wasn't able to provoke a test failure by adding unused storage variables to the test contracts. Even after removing the ``mstore`` call all tests passed. Now I'm a bit scared ;) Do you have an idea how we could bisect this? Doesn't ``keccak`` expects the array reference to be stored in memory?

Reply:
Probably the most reliable way to create such a test is to just use `assembly { mstore(0, "garbage") } x.pop();`